assignments/prototypes.js

```diff
@@ -32,79 +54,189 @@
+console.log(mage.takeDamage()); // Bruce took damage.
+console.log(swordsman.destroy()); // Sir Mustachio was removed from the game.
+
+// Stretch task:
+// * Create Villain and Hero constructor functions that inherit from the Humanoid constructor function.
+// * Give the Hero and Villains different methods that could be used to remove health points from objects which could result in destruction if health gets to 0 or drops below 0;
+// * Create two new objects, one a villain and one a hero and fight it out with methods!
+
+function Character(attributes) {
+  Humanoid.call(this, attributes);
+  this.ac = attributes.ac;
+  this.initiative = attributes.initiative;
+  this.attack = attributes.attack;
+  this.damage = attributes.damage;
+  this.damageDie = attributes.damageDie;
+}
+
+//Roll dice by passing in number of sides and an optional bonus to the roll
+function rollDice(numSides, bonus = 0) {
+  return Math.floor(Math.random() * (numSides + 1) + bonus);
+}
+
+Character.prototype = Object.create(Humanoid.prototype);
+
+//Characters attacking a specified target
+Character.prototype.doAttack = function(target) {
+  //Check to see if the attack roll meets the target's AC
+  if (rollDice(20, this.attack) >= target.ac) {
+    //If so, we roll damage
+    target.healthPoints -= rollDice(this.damageDie, this.damage);
+    console.log(target.takeDamage());
+
+    if (target.healthPoints <= 0) {
+      console.log(target.destroy());
+    }
+  } else {
+    console.log(`${this.name} misses!`);
+  }
+};
+
+//Characters rolling initiative (turn order)
+Character.prototype.rollInitiative = function() {
+  return rollDice(20, this.initiative);
+};
+
+const hero = new Character({
+  createdAt: new Date(),
+  dimensions: {
+    length: 2,
+    width: 2,
+    height: 4,
+  },
+  healthPoints: 55,
+  name: 'Jason',
+  team: 'Knights of the Crucible',
+  weapons: ['Sword', 'Bow'],
+  language: 'Common',
+  initiative: 3,
+  ac: 22,
+  attack: 7,
+  damage: 4,
+  damageDie: 12,
+});
+
+const villain = new Character({
+  createdAt: new Date(),
+  dimensions: {
+    length: 3,
+    width: 4,
+    height: 6,
+  },
+  healthPoints: 74,
+  name: 'Belphegor',
+  team: 'Demon Army',
+  weapons: ['Dark Claw'],
+  language: 'Demonic',
+  initiative: 2,
+  ac: 16,
+  attack: 8,
+  damage: 6,
+  damageDie: 8,
+});
+
+//Using closure to set up a basic turn counter function
+const turnCounter = () => {
+  let counter = 0;
+  return () => {
+    counter++;
+    console.log(`Turn ${counter}`);
+  };
+};
+
+//Combat logic
+function doCombat(person1, person2) {
+  //Compare initiative rolls and determine turn order
+  let first =
+    person1.rollInitiative() >= person2.rollInitiative() ? person1 : person2;
+  let second = first === person1 ? person2 : person1;
+  const turns = turnCounter();
+
+  //Continue as long as both combatants have health left
+  while (person1.healthPoints > 0 && person2.healthPoints > 0) {
```

Comment:
Using a while loop here to have a mock battle was a great idea, and using things that haven't been covered is amazing to see.